polygon/heimdall/entity_fetcher.go

```diff
@@ -0,0 +1,133 @@
+package heimdall
+
+import (
+	"cmp"
+	"context"
+	"fmt"
+	"slices"
+	"time"
+
+	"github.com/ledgerwatch/log/v3"
+)
+
+type entityFetcher interface {
+	FetchLastEntityId(ctx context.Context) (uint64, error)
+	FetchEntitiesRange(ctx context.Context, idRange EntityIdRange) ([]any, error)
+}
+
+type genericEntityFetcher[TEntity entity] struct {
+	name string
+
+	fetchLastEntityId func(ctx context.Context) (int64, error)
+	fetchEntity       func(ctx context.Context, id int64) (*TEntity, error)
+	fetchEntitiesPage func(ctx context.Context, page uint64, limit uint64) ([]*TEntity, error)
+	getStartBlockNum  func(entity *TEntity) uint64
+
+	logger log.Logger
+}
+
+func newGenericEntityFetcher[TEntity entity](
+	name string,
+	fetchLastEntityId func(ctx context.Context) (int64, error),
+	fetchEntity func(ctx context.Context, id int64) (*TEntity, error),
+	fetchEntitiesPage func(ctx context.Context, page uint64, limit uint64) ([]*TEntity, error),
+	getStartBlockNum func(entity *TEntity) uint64,
+	logger log.Logger,
+) entityFetcher {
+	return &genericEntityFetcher[TEntity]{
+		name:              name,
+		fetchLastEntityId: fetchLastEntityId,
+		fetchEntity:       fetchEntity,
+		fetchEntitiesPage: fetchEntitiesPage,
+		getStartBlockNum:  getStartBlockNum,
+		logger:            logger,
+	}
+}
+
+func (f *genericEntityFetcher[TEntity]) FetchLastEntityId(ctx context.Context) (uint64, error) {
+	id, err := f.fetchLastEntityId(ctx)
+	return uint64(id), err
+}
+
+func (f *genericEntityFetcher[TEntity]) FetchEntitiesRange(ctx context.Context, idRange EntityIdRange) ([]any, error) {
+	count := idRange.Len()
+
+	const batchFetchThreshold = 100
+	if (count > batchFetchThreshold) && (f.fetchEntitiesPage != nil) {
+		allEntities, err := f.FetchAllEntities(ctx)
+		if err != nil {
+			return nil, err
+		}
+		startIndex := idRange.Start - 1
+		return allEntities[startIndex : startIndex+count], nil
+	}
+
+	return f.FetchEntitiesRangeSequentially(ctx, idRange)
+}
+
+func (f *genericEntityFetcher[TEntity]) FetchEntitiesRangeSequentially(ctx context.Context, idRange EntityIdRange) ([]any, error) {
+	entities := make([]any, 0, idRange.Len())
+
+	for id := idRange.Start; id <= idRange.End; id++ {
+		entity, err := f.fetchEntity(ctx, int64(id))
+		if err != nil {
+			return nil, err
+		}
+		entities = append(entities, entity)
+	}
+
+	return entities, nil
+}
+
+func (f *genericEntityFetcher[TEntity]) FetchAllEntities(ctx context.Context) ([]any, error) {
+	// TODO: once heimdall API is fixed to return sorted items in pages we can only fetch
+	//
+	//	the new pages after lastStoredCheckpointId using the checkpoints/list paging API
+	//	(for now we have to fetch all of them)
+	//	and also remove sorting we do after fetching
+
+	var entities []any
+
+	fetchStartTime := time.Now()
+	progressLogTicker := time.NewTicker(30 * time.Second)
+	defer progressLogTicker.Stop()
+
+	for page := uint64(1); page < 10_000; page++ {
```

Comment:
why "page < 10_000"?
the logic here looks incorrect:
- first you need to fetch the count of the entities by using the count API
- then you need to fetch them all using paging - please look at how this is implemented in `batchFetchCheckpoints` with `for count >0`

Reply:
This logic is already adapted from batchFetchCheckpoints.
The logic is correct. I've checked that requesting the last + 1 page returns an empty list, then I exit the loop.

I didn't want an endless loop, so I've added this arbitrary 10K constant (could be `uint64.Max`). I see now that this is stupid, because it will not happen in practice, and just confuses people. I can simply remove this `page < 10_000` to save confusion.

Reply:
removing the condition sounds good to me